consensus/view_change_construct.go

```diff
@@ -0,0 +1,422 @@
+
+	vc.m1Payload = []byte{}
+	vc.bhpSigs = map[uint64]map[string]*bls_core.Sign{}
+	vc.nilSigs = map[uint64]map[string]*bls_core.Sign{}
+	vc.viewIDSigs = map[uint64]map[string]*bls_core.Sign{}
+	vc.bhpBitmap = map[uint64]*bls_cosi.Mask{}
+	vc.nilBitmap = map[uint64]*bls_cosi.Mask{}
+	vc.viewIDBitmap = map[uint64]*bls_cosi.Mask{}
+	vc.newViewMsg = map[uint64]map[string]uint64{}
+}
+
+// GetPreparedBlock returns the prepared block or nil if not found
+func (vc *viewChange) GetPreparedBlock(fbftlog *FBFTLog, hash [32]byte) ([]byte, []byte) {
+	vc.vcLock.RLock()
+	defer vc.vcLock.RUnlock()
+
+	if !vc.isM1PayloadEmpty() {
+		block := fbftlog.GetBlockByHash(hash)
+		if block != nil {
+			encodedBlock, err := rlp.EncodeToBytes(block)
+			if err != nil || len(encodedBlock) == 0 {
+				vc.getLogger().Err(err).Msg("[GetPreparedBlock] Failed encoding prepared block")
+				return vc.m1Payload, nil
+			}
+			return vc.m1Payload, encodedBlock
+		}
+	}
+	return vc.m1Payload, nil
+}
+
+// GetM2Bitmap returns the nilBitmap as M2Bitmap
+func (vc *viewChange) GetM2Bitmap(viewID uint64) ([]byte, []byte) {
+	vc.vcLock.RLock()
+	defer vc.vcLock.RUnlock()
+
+	sig2arr := []*bls_core.Sign{}
+	for _, sig := range vc.nilSigs[viewID] {
+		sig2arr = append(sig2arr, sig)
+	}
+
+	if len(sig2arr) > 0 {
+		m2Sig := bls_cosi.AggregateSig(sig2arr)
+		vc.getLogger().Info().Int("len", len(sig2arr)).Msg("[GetM2Bitmap] M2 (NIL) type signatures")
+		return m2Sig.Serialize(), vc.nilBitmap[viewID].Bitmap
+	}
+	vc.getLogger().Info().Uint64("viewID", viewID).Msg("[GetM2Bitmap] No M2 (NIL) type signatures")
+	return nil, nil
+}
+
+// GetM3Bitmap returns the viewIDBitmap as M3Bitmap
+func (vc *viewChange) GetM3Bitmap(viewID uint64) ([]byte, []byte) {
+	vc.vcLock.RLock()
+	defer vc.vcLock.RUnlock()
+
+	sig3arr := []*bls_core.Sign{}
+	for _, sig := range vc.viewIDSigs[viewID] {
+		sig3arr = append(sig3arr, sig)
+	}
+	// even we check here for safty, m3 type signatures must >= 2f+1
+	if len(sig3arr) > 0 {
+		m3Sig := bls_cosi.AggregateSig(sig3arr)
+		vc.getLogger().Info().Int("len", len(sig3arr)).Msg("[GetM3Bitmap] M3 (ViewID) type signatures")
+		return m3Sig.Serialize(), vc.viewIDBitmap[viewID].Bitmap
+	}
+	vc.getLogger().Info().Uint64("viewID", viewID).Msg("[GetM3Bitmap] No M3 (ViewID) type signatures")
+	return nil, nil
+}
+
+// VerifyNewViewMsg returns true if the new view message is valid
+func (vc *viewChange) VerifyNewViewMsg(recvMsg *FBFTMessage) (*types.Block, error) {
+	vc.vcLock.Lock()
+	defer vc.vcLock.Unlock()
+
+	if recvMsg.M3AggSig == nil || recvMsg.M3Bitmap == nil {
+		return nil, errors.New("[VerifyNewViewMsg] M3AggSig or M3Bitmap is nil")
+	}
+
+	senderKey := recvMsg.SenderPubkey
+	senderKeyStr := senderKey.Bytes.Hex()
+
+	// first time received the new view message
+	_, ok := vc.newViewMsg[recvMsg.ViewID]
+	if !ok {
+		newViewMap := map[string]uint64{}
+		vc.newViewMsg[recvMsg.ViewID] = newViewMap
+	}
+
+	_, ok = vc.newViewMsg[recvMsg.ViewID][senderKeyStr]
```

Comment:
Don't think there is a need to add this newViewMsg map and check whether new view is already received. As long as the newView message is verified and valid, we should always response to it. Cases, like the leader failed to deliver newView message to everyone and need retry logic, in this case, better let everyone to always response to the same newView message until consensus can be reached.

Reply:
this situation won't happen as the validator received a valid newview message, it will enter normal mode, not in viewchange anymore. So, it won't process the newview message. Otherwise, it's a security risk.  

The original function locked the entire function, so there is not reentry of the processing logic.  Now, since I removed the monolithic lock, I simply don't want to process the same message again, just like how we handle viewchange messages.

Reply:
I think we can err on the safer side and respond to the same message as long as it's valid (meaning I am still in view change, and block number and all other checks pass). Of course, we don't need to respond to a newView message if we already enter normal mode. But before we successfully finish a view change, better respond to every messages for better robustness. Otherwise, think about cases, where half of the nodes responds but others don't. Then if the leader is restarted, it will lose the collected votes from the first half and when it retries, those first half won't respond anymore and this view change won't succeed anymore on retries.